### PR TITLE
Lint errors

### PR DESCRIPTION
**Pull request by threepointone**
This PR fixes a number of no-undef errors in /lib and /test.

It also introduces babel and eslint as tools for linting / compiling, and sample .babelrc/.eslintrc files.  

(This PR also includes commits from #48)

```
eslint lib --quiet 
✖ 157 problems (157 errors, 0 warnings)

eslint test --quiet 
✖ 9 problems (9 errors, 0 warnings)

/Users/sunilpai/code/reqlite/lib/changes.js
  105:34  error  Missing whitespace after semicolon  semi-spacing

/Users/sunilpai/code/reqlite/lib/date.js
  18:14  error  "ReqlGeometry" is not defined  no-undef

/Users/sunilpai/code/reqlite/lib/geo.js
  41:92  error  "query" is not defined  no-undef

/Users/sunilpai/code/reqlite/lib/maxval.js
  9:73  error  "query" is not defined  no-undef

/Users/sunilpai/code/reqlite/lib/minval.js
  9:73  error  "query" is not defined  no-undef

/Users/sunilpai/code/reqlite/lib/node.js
   4:1   error  "options" is not defined  no-undef
   8:7   error  "options" is not defined  no-undef
  11:7   error  "options" is not defined  no-undef
  16:7   error  "options" is not defined  no-undef
  19:25  error  "options" is not defined  no-undef

/Users/sunilpai/code/reqlite/lib/query.js
    65:7   error  "toReturn" is not defined           no-undef
    68:9   error  "toReturn" is not defined           no-undef
    70:7   error  "toReturn" is not defined           no-undef
    70:31  error  "toReturn" is not defined           no-undef
    73:12  error  "toReturn" is not defined           no-undef
   693:38  error  "message" is not defined            no-undef
   772:5   error  Missing whitespace after semicolon  semi-spacing
  1020:11  error  "reference" is not defined          no-undef
  1022:28  error  "reference" is not defined          no-undef
  1023:27  error  "reference" is not defined          no-undef
  1102:21  error  "sequence" is not defined           no-undef
  1121:21  error  "sequence" is not defined           no-undef
  1133:52  error  Missing whitespace after semicolon  semi-spacing
  1532:7   error  Missing whitespace after semicolon  semi-spacing
  1619:5   error  Missing whitespace after semicolon  semi-spacing
  1988:5   error  "value" is not defined              no-undef
  1989:35  error  "value" is not defined              no-undef
  1992:25  error  "value" is not defined              no-undef
  1995:11  error  "value" is not defined              no-undef
  1999:16  error  "value" is not defined              no-undef
  2002:31  error  "value" is not defined              no-undef
  2004:32  error  "value" is not defined              no-undef
  2005:44  error  "value" is not defined              no-undef
  2007:30  error  "value" is not defined              no-undef
  2008:27  error  "value" is not defined              no-undef
  2010:35  error  "value" is not defined              no-undef
  2011:44  error  "value" is not defined              no-undef
  2017:27  error  "value" is not defined              no-undef
  2018:16  error  "value" is not defined              no-undef
  2020:30  error  "value" is not defined              no-undef
  2024:35  error  "value" is not defined              no-undef
  2026:32  error  "value" is not defined              no-undef
  2030:21  error  "value" is not defined              no-undef
  2052:27  error  "value" is not defined              no-undef
  2056:24  error  "value" is not defined              no-undef
  2057:18  error  "value" is not defined              no-undef
  2066:30  error  "value" is not defined              no-undef
  2070:35  error  "value" is not defined              no-undef
  2071:16  error  "value" is not defined              no-undef
  2080:28  error  "value" is not defined              no-undef
  2083:30  error  "value" is not defined              no-undef
  2084:16  error  "value" is not defined              no-undef
  2358:7   error  "dbProvided" is not defined         no-undef
  2530:19  error  "error" is not defined              no-undef
  2573:7   error  "bool" is not defined               no-undef
  2574:22  error  "index" is not defined              no-undef
  2575:21  error  "bool" is not defined               no-undef
  2577:21  error  "bool" is not defined               no-undef
  2578:14  error  "bool" is not defined               no-undef
  2581:7   error  "left" is not defined               no-undef
  2582:22  error  "index" is not defined              no-undef
  2583:19  error  "left" is not defined               no-undef
  2585:5   error  "bool" is not defined               no-undef
  2588:28  error  "bool" is not defined               no-undef
  2589:28  error  "left" is not defined               no-undef
  2589:34  error  "index" is not defined              no-undef
  2918:5   error  "momentDate" is not defined         no-undef
  2919:9   error  "momentDate" is not defined         no-undef
  2920:14  error  "momentDate" is not defined         no-undef
  2920:35  error  "momentDate" is not defined         no-undef
  2923:14  error  "momentDate" is not defined         no-undef
  2923:35  error  "momentDate" is not defined         no-undef
  3018:5   error  "momentDate" is not defined         no-undef
  3020:5   error  "momentDate" is not defined         no-undef
  3020:25  error  "momentDate" is not defined         no-undef
  3021:5   error  "momentDate" is not defined         no-undef
  3021:25  error  "momentDate" is not defined         no-undef
  3022:5   error  "momentDate" is not defined         no-undef
  3022:25  error  "momentDate" is not defined         no-undef
  3023:5   error  "momentDate" is not defined         no-undef
  3023:25  error  "momentDate" is not defined         no-undef
  3025:32  error  "momentDate" is not defined         no-undef
  3036:5   error  "momentDate" is not defined         no-undef
  3037:18  error  "momentDate" is not defined         no-undef
  3038:15  error  "momentDate" is not defined         no-undef
  3039:15  error  "momentDate" is not defined         no-undef
  3040:15  error  "momentDate" is not defined         no-undef
  3063:5   error  "momentDate" is not defined         no-undef
  3064:12  error  "momentDate" is not defined         no-undef
  3089:5   error  "momentDate" is not defined         no-undef
  3090:12  error  "momentDate" is not defined         no-undef
  3101:5   error  "momentDate" is not defined         no-undef
  3102:12  error  "momentDate" is not defined         no-undef
  3113:5   error  "momentDate" is not defined         no-undef
  3114:12  error  "momentDate" is not defined         no-undef
  3125:5   error  "momentDate" is not defined         no-undef
  3126:12  error  "momentDate" is not defined         no-undef
  3137:5   error  "momentDate" is not defined         no-undef
  3138:12  error  "momentDate" is not defined         no-undef
  3149:5   error  "momentDate" is not defined         no-undef
  3150:12  error  "momentDate" is not defined         no-undef
  3150:33  error  "momentDate" is not defined         no-undef
  3406:13  error  "result" is not defined             no-undef
  3407:25  error  "result" is not defined             no-undef
  3411:17  error  "result" is not defined             no-undef
  3411:26  error  "result" is not defined             no-undef
  3412:32  error  "result" is not defined             no-undef
  3413:47  error  "result" is not defined             no-undef
  3416:17  error  "result" is not defined             no-undef
  3416:26  error  "result" is not defined             no-undef
  3419:17  error  "result" is not defined             no-undef
  3419:26  error  "result" is not defined             no-undef
  3419:56  error  "result" is not defined             no-undef
  3422:33  error  "result" is not defined             no-undef
  3504:22  error  "max" is not defined                no-undef

/Users/sunilpai/code/reqlite/lib/sequence.js
    11:37  error  Missing whitespace after semicolon  semi-spacing
   518:14  error  "result" is not defined             no-undef
   779:9   error  "varIds" is not defined             no-undef
   780:23  error  "varIds" is not defined             no-undef
   781:23  error  "varIds" is not defined             no-undef
   791:28  error  "varIds" is not defined             no-undef
   797:32  error  "varIds" is not defined             no-undef
   798:32  error  "varIds" is not defined             no-undef
  1152:11  error  "fnResult" is not defined           no-undef
  1153:39  error  "fnResult" is not defined           no-undef
  1154:11  error  "fnResult" is not defined           no-undef
  1154:39  error  "fnResult" is not defined           no-undef
  1156:18  error  "fnResult" is not defined           no-undef

/Users/sunilpai/code/reqlite/lib/table.js
  118:5   error  "typeValue" is not defined  no-undef
  119:56  error  "typeValue" is not defined  no-undef
  127:7   error  "newResult" is not defined  no-undef
  128:37  error  "newResult" is not defined  no-undef
  132:5   error  "newResult" is not defined  no-undef
  133:35  error  "newResult" is not defined  no-undef
  573:7   error  "fieldOrFn" is not defined  no-undef
  575:33  error  "fieldOrFn" is not defined  no-undef
  577:32  error  "fieldOrFn" is not defined  no-undef
  681:5   error  "fn" is not defined         no-undef
  682:31  error  "fn" is not defined         no-undef
  693:29  error  "fn" is not defined         no-undef

/Users/sunilpai/code/reqlite/lib/utils/main.js
   718:7   error  "raw_date_str" is not defined  no-undef
   721:7   error  "raw_date_str" is not defined  no-undef
   725:5   error  "raw_date_str" is not defined  no-undef
   729:10  error  "raw_date_str" is not defined  no-undef
   729:32  error  "raw_date_str" is not defined  no-undef
   821:9   error  "righ" is not defined          no-undef
  1051:61  error  "field" is not defined         no-undef
  1354:3   error  "keep" is not defined          no-undef
  1357:7   error  "keep" is not defined          no-undef
  1362:7   error  "keep" is not defined          no-undef
  1368:7   error  "keep" is not defined          no-undef
  1373:7   error  "keep" is not defined          no-undef
  1377:7   error  "keep" is not defined          no-undef
  1506:5   error  "angle" is not defined         no-undef
  1507:63  error  "angle" is not defined         no-undef
  1582:15  error  "value" is not defined         no-undef
  1611:17  error  "value" is not defined         no-undef



/Users/sunilpai/code/reqlite/test/dates-and-times.js
  54:7  error  Missing whitespace after semicolon  semi-spacing

/Users/sunilpai/code/reqlite/test/geo.js
  1048:111  error  Missing whitespace after semicolon  semi-spacing
  1053:129  error  Missing whitespace after semicolon  semi-spacing
  1063:93   error  Missing whitespace after semicolon  semi-spacing
  1068:112  error  Missing whitespace after semicolon  semi-spacing
  1073:101  error  Missing whitespace after semicolon  semi-spacing

/Users/sunilpai/code/reqlite/test/manipulating-databases.js
  84:44  error  Unexpected whitespace before semicolon  semi-spacing

/Users/sunilpai/code/reqlite/test/manipulating-tables.js
  120:47  error  Unexpected whitespace before semicolon  semi-spacing
  139:47  error  Unexpected whitespace before semicolon  semi-spacing
```

### CODE REVIEW

**Comment by neumino** [comment]
What's the reason behind adding babel? I don't think we need it.

I don't like lint services for multiple reasons, but I'll make an exception for reqlite since it's new-ish and it actually catches a few bugs.

A few things though:
- Code that is obviously wrong should be fixed, not commented
- Dead code should be removed, not commented
- Add tests for whatever is broken if possible. It's fairly easy to do as you just have to write a query.
- No spacing between operators

I can help for the first three points.

**Comment by threepointone** [comment]
Will remove Babel, was more habit than anything. I'll also try to incorporate rules for the rest.

File: lib/geo.js [hunk 1]
```diff
@@ -38,11 +38,11 @@ ReqlGeometry.prototype.distance = function(to, options) {
 
   if (to.type !== 'Point') {
     //TODO Implement
-    util.notAvailable('You can compute the distance to a point only in Reqlite (for now)', query)
+    util.notAvailable('You can compute the distance to a point only in Reqlite (for now)', /* ?? query */ {})
```
Comment (threepointone) [inline lib/geo.js]: is this right? there's no access to a `query` object in this scope, so I sent a regular object instead.

File: lib/maxval.js [hunk 2]
```diff
@@ -3,9 +3,9 @@ function Maxval() {
 }
 module.exports = Maxval;
 
-var Error = require(__dirname+"/error.js");
+var Error = require("./error.js");
 
 Maxval.prototype.toJSON = function() {
-  throw new Error.ReqlRuntimeError('Cannot convert `r.maxval` to JSON', query.frames)
+  throw new Error.ReqlRuntimeError('Cannot convert `r.maxval` to JSON' /* ?? , query.frames */)
```
Comment (threepointone) [inline lib/maxval.js]: no query available here either

File: lib/minval.js [hunk 3]
```diff
@@ -3,8 +3,8 @@ function Minval() {
 }
 module.exports = Minval;
 
-var Error = require(__dirname+"/error.js");
+var Error = require("./error.js");
 
 Minval.prototype.toJSON = function() {
-  throw new Error.ReqlRuntimeError('Cannot convert `r.minval` to JSON', query.frames)
+  throw new Error.ReqlRuntimeError('Cannot convert `r.minval` to JSON' /* ?? , query.frames */)
```
Comment (threepointone) [inline lib/minval.js]: ditto

File: lib/query.js [hunk 4]
```diff
@@ -690,7 +691,7 @@ Query.prototype.error = function(args, options, internalOptions) {
   self.frames.push(0);
   if (args.length === 0) {
     //TODO Add test
-    throw new Error.ReqlRuntimeError(message, self.frames);
+    throw new Error.ReqlRuntimeError('Query.prototype.error!', self.frames);
```
Comment (threepointone) [inline lib/query.js]: any idea for a better message?

File: lib/table.js [hunk 5]
```diff
@@ -124,13 +124,13 @@ Table.prototype.insert = function(docs, options, query) {
   var newDoc;
   if (util.isSequence(docs)) {
     for(var i=0; i<docs.length; i++) {
-      newResult = this._singleInsert(docs.get(i), options, query);
-      util.mergeWriteResult(result, newResult);
+      newDoc = this._singleInsert(docs.get(i), options, query);
```
Comment (threepointone) [inline lib/table.js]: there's an undefined/ununsed variable here `generatedKeys` that seems important?

**Comment by threepointone** [comment]
Do you plan on using es6/+ features on node? fat arrows etc.

**Comment by threepointone** [comment]
removed babel, added an npm command `npm run lint` to check for critical errors (should I run that in `npm test` too?)

File: lib/geo.js (hunk 1 above)
Comment (neumino) [inline lib/geo.js]: The query should be passed in the argument of `distance`

File: lib/maxval.js (hunk 2 above)
Comment (neumino) [inline lib/maxval.js]: This is a bit tricky to fix, we can just remove the query for now.
To properly fix it, we need to add a new internal option to determine if `r.asc/desc` is valid in the given context.

File: lib/query.js (hunk 4 above)
Comment (neumino) [inline lib/query.js]: It should match what RethinkDB is sending. Not sure what it's sending though.

File: lib/table.js (hunk 5 above)
Comment (neumino) [inline lib/table.js]: Yes, it's a new variable, we shouldn't reuse newDoc

**Comment by neumino** [comment]
No plan to use es6 features for now.
Can you put lint in a new command and in test too?

**Comment by neumino** [comment]
Merging it, I'll update a few more things before pushing.

**Comment by neumino** [comment]
It's in the branch `next`. It's taking a bit longer than expected to properly clean things, but it's getting there (slowly).

**Comment by threepointone** [comment]
Neat! I'll pitch in later today, we'll do this together.

**Pull request merged**